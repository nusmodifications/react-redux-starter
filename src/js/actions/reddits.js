// @flow
import type { FSA } from 'types/redux';
import { API_REQUEST } from 'middlewares/requests-middleware';

export const FETCH_REDDITS: string = 'FETCH_REDDITS';
export function fetchReddits(topic: string): FSA {
  return {
    type: FETCH_REDDITS,
    payload: {
      method: 'GET',
      url: `https://www.reddit.com/r/${topic}.json`,
    },
    meta: {
      [API_REQUEST]: true,
    },
  };
}
