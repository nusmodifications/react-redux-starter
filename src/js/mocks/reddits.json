{
  "kind": "Listing",
  "data": {
    "modhash": "",
    "children": [{
      "kind": "t3",
      "data": {
        "contest_mode": false,
        "banned_by": null,
        "media_embed": {},
        "subreddit": "reactjs",
        "selftext_html": null,
        "selftext": "",
        "likes": null,
        "suggested_sort": null,
        "user_reports": [],
        "secure_media": null,
        "link_flair_text": null,
        "id": "5v417g",
        "gilded": 0,
        "secure_media_embed": {},
        "clicked": false,
        "score": 15,
        "report_reasons": null,
        "author": "ummahusla",
        "saved": false,
        "mod_reports": [],
        "name": "t3_5v417g",
        "subreddit_name_prefixed": "r/reactjs",
        "approved_by": null,
        "over_18": false,
        "domain": "twitter.com",
        "hidden": false,
        "preview": {
          "images": [{
            "source": {
              "url": "https://i.redditmedia.com/-CfDEf2LPRvQz8WpCmiDNh9pIphAwmgmvYl6q_tXI0c.jpg?s=79001baa363348139b8014e87835411c",
              "width": 400,
              "height": 400
            },
            "resolutions": [{
              "url": "https://i.redditmedia.com/-CfDEf2LPRvQz8WpCmiDNh9pIphAwmgmvYl6q_tXI0c.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=108&amp;s=a29921fcad4a0acd4a09628232957d6b",
              "width": 108,
              "height": 108
            }, {
              "url": "https://i.redditmedia.com/-CfDEf2LPRvQz8WpCmiDNh9pIphAwmgmvYl6q_tXI0c.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=216&amp;s=e2d015629c7402c3b0a33e86ffb49d37",
              "width": 216,
              "height": 216
            }, {
              "url": "https://i.redditmedia.com/-CfDEf2LPRvQz8WpCmiDNh9pIphAwmgmvYl6q_tXI0c.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=320&amp;s=7e4f73c6faac946709e12fc30e99566f",
              "width": 320,
              "height": 320
            }],
            "variants": {},
            "id": "asSC2vPGcuO1OW5zJoMSc_HrdNXODaPVuIxxwfhBmiU"
          }],
          "enabled": false
        },
        "thumbnail": "http://b.thumbs.redditmedia.com/BGAn4ep7OtJlGWIquzdTnSeN5jyIU57PFdCxes21S2g.jpg",
        "subreddit_id": "t5_2zldd",
        "edited": false,
        "link_flair_css_class": null,
        "author_flair_css_class": null,
        "downs": 0,
        "brand_safe": true,
        "archived": false,
        "removal_reason": null,
        "post_hint": "link",
        "is_self": false,
        "hide_score": false,
        "spoiler": false,
        "permalink": "/r/reactjs/comments/5v417g/react_has_60000_github_stars/",
        "num_reports": null,
        "locked": false,
        "stickied": false,
        "created": 1487618590.0,
        "url": "https://twitter.com/reactjs/status/833570245243654144",
        "author_flair_text": null,
        "quarantine": false,
        "title": "React has 60,000 GitHub stars",
        "created_utc": 1487589790.0,
        "distinguished": null,
        "media": null,
        "num_comments": 6,
        "visited": false,
        "subreddit_type": "public",
        "ups": 15
      }
    }, {
      "kind": "t3",
      "data": {
        "contest_mode": false,
        "banned_by": null,
        "media_embed": {},
        "subreddit": "reactjs",
        "selftext_html": null,
        "selftext": "",
        "likes": null,
        "suggested_sort": null,
        "user_reports": [],
        "secure_media": null,
        "link_flair_text": null,
        "id": "5v4t16",
        "gilded": 0,
        "secure_media_embed": {},
        "clicked": false,
        "score": 3,
        "report_reasons": null,
        "author": "LikwidSkin",
        "saved": false,
        "mod_reports": [],
        "name": "t3_5v4t16",
        "subreddit_name_prefixed": "r/reactjs",
        "approved_by": null,
        "over_18": false,
        "domain": "medium.com",
        "hidden": false,
        "preview": {
          "images": [{
            "source": {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?s=22a96195e04dbefe52e70dfa717442a0",
              "width": 1200,
              "height": 258
            },
            "resolutions": [{
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=108&amp;s=8214acea8b1f6bc1eba5f19528d1fe6c",
              "width": 108,
              "height": 23
            }, {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=216&amp;s=832b3a6c9f5f6e70a26307500f2827ac",
              "width": 216,
              "height": 46
            }, {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=320&amp;s=6433397e3bbc0f995df64ab43fbb11df",
              "width": 320,
              "height": 68
            }, {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=640&amp;s=d613fe54a905087922f796ddeac86e5a",
              "width": 640,
              "height": 137
            }, {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=960&amp;s=c17b230e1b063bae4ce8760b7c4ea919",
              "width": 960,
              "height": 206
            }, {
              "url": "https://i.redditmedia.com/bOZxa3Eijn-DBmib-gg4pG1tbcqWbgcePOdjdfC_ykc.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=1080&amp;s=d96a05ffe3954562192e69b47fbf745f",
              "width": 1080,
              "height": 232
            }],
            "variants": {},
            "id": "4PkDA-aOPWxk5sTeY8PZFuBDkArtDoNyzDqwL5Gmwbo"
          }],
          "enabled": false
        },
        "thumbnail": "http://b.thumbs.redditmedia.com/9H6XPpNQpY4pCnHoCImwUKmiUk81wwBAFpxyKJoHR_M.jpg",
        "subreddit_id": "t5_2zldd",
        "edited": false,
        "link_flair_css_class": null,
        "author_flair_css_class": null,
        "downs": 0,
        "brand_safe": true,
        "archived": false,
        "removal_reason": null,
        "post_hint": "link",
        "is_self": false,
        "hide_score": true,
        "spoiler": false,
        "permalink": "/r/reactjs/comments/5v4t16/reactjs_to_reduce_load_time_by_250_in_a_month/",
        "num_reports": null,
        "locked": false,
        "stickied": false,
        "created": 1487629786.0,
        "url": "https://medium.com/urbanclap-engineering/improving-first-time-load-of-a-production-react-app-part-1-of-2-e7494a7c7ab0#.ht05owhba",
        "author_flair_text": null,
        "quarantine": false,
        "title": "ReactJs to reduce load time by 250% in a month",
        "created_utc": 1487600986.0,
        "distinguished": null,
        "media": null,
        "num_comments": 0,
        "visited": false,
        "subreddit_type": "public",
        "ups": 3
      }
    }, {
      "kind": "t3",
      "data": {
        "contest_mode": false,
        "banned_by": null,
        "media_embed": {},
        "subreddit": "reactjs",
        "selftext_html": null,
        "selftext": "",
        "likes": null,
        "suggested_sort": null,
        "user_reports": [],
        "secure_media": null,
        "link_flair_text": null,
        "id": "5v50hp",
        "gilded": 0,
        "secure_media_embed": {},
        "clicked": false,
        "score": 2,
        "report_reasons": null,
        "author": "WeLikeGraphQL",
        "saved": false,
        "mod_reports": [],
        "name": "t3_5v50hp",
        "subreddit_name_prefixed": "r/reactjs",
        "approved_by": null,
        "over_18": false,
        "domain": "medium.com",
        "hidden": false,
        "thumbnail": "default",
        "subreddit_id": "t5_2zldd",
        "edited": false,
        "link_flair_css_class": null,
        "author_flair_css_class": null,
        "downs": 0,
        "brand_safe": true,
        "archived": false,
        "removal_reason": null,
        "is_self": false,
        "hide_score": true,
        "spoiler": false,
        "permalink": "/r/reactjs/comments/5v50hp/use_of_recompose_in_universal_react_apollo_example/",
        "num_reports": null,
        "locked": false,
        "stickied": false,
        "created": 1487632255.0,
        "url": "https://medium.com/welikegraphql/use-of-recompose-in-universal-react-apollo-example-3d1f89bc945b#.ilom8jbia",
        "author_flair_text": null,
        "quarantine": false,
        "title": "Use of Recompose in Universal React Apollo Example",
        "created_utc": 1487603455.0,
        "distinguished": null,
        "media": null,
        "num_comments": 0,
        "visited": false,
        "subreddit_type": "public",
        "ups": 2
      }
    }, {
      "kind": "t3",
      "data": {
        "contest_mode": false,
        "banned_by": null,
        "media_embed": {},
        "subreddit": "reactjs",
        "selftext_html": null,
        "selftext": "",
        "likes": null,
        "suggested_sort": null,
        "user_reports": [],
        "secure_media": null,
        "link_flair_text": null,
        "id": "5v24pt",
        "gilded": 0,
        "secure_media_embed": {},
        "clicked": false,
        "score": 12,
        "report_reasons": null,
        "author": "logandice",
        "saved": false,
        "mod_reports": [],
        "name": "t3_5v24pt",
        "subreddit_name_prefixed": "r/reactjs",
        "approved_by": null,
        "over_18": false,
        "domain": "webacademy.io",
        "hidden": false,
        "preview": {
          "images": [{
            "source": {
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?s=5e5a20d8f5091ee94f40c9a9291150ce",
              "width": 960,
              "height": 543
            },
            "resolutions": [{
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=108&amp;s=d5981382a31714b76b679e0a3b77e142",
              "width": 108,
              "height": 61
            }, {
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=216&amp;s=5dbfff87f27e6e704523dbb749c6f4f5",
              "width": 216,
              "height": 122
            }, {
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=320&amp;s=9a06a49e484282a18b0789c9763bb20a",
              "width": 320,
              "height": 181
            }, {
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=640&amp;s=73f18cd22c8c3923ffbe584af43eba64",
              "width": 640,
              "height": 362
            }, {
              "url": "https://i.redditmedia.com/aDBZTmhkfHZ9KzMk3ptRLbTyCuMbAdA7FrIC4TDZCIw.jpg?fit=crop&amp;crop=faces%2Centropy&amp;arh=2&amp;w=960&amp;s=3e7a218c1bcf2a168b3aeb1553360a23",
              "width": 960,
              "height": 543
            }],
            "variants": {},
            "id": "mL6Qvr75m670uqc9vA72ARnMG5GV11cUTLmu95Lp7ZY"
          }],
          "enabled": false
        },
        "thumbnail": "http://b.thumbs.redditmedia.com/DozvMWDJ4Z7s4aEt1pIh-fJPXxysXQFnqGUNSrxSH0w.jpg",
        "subreddit_id": "t5_2zldd",
        "edited": false,
        "link_flair_css_class": null,
        "author_flair_css_class": null,
        "downs": 0,
        "brand_safe": true,
        "archived": false,
        "removal_reason": null,
        "post_hint": "link",
        "is_self": false,
        "hide_score": false,
        "spoiler": false,
        "permalink": "/r/reactjs/comments/5v24pt/best_react_tutorials_and_courses/",
        "num_reports": null,
        "locked": false,
        "stickied": false,
        "created": 1487587715.0,
        "url": "http://webacademy.io/react-tutorials/",
        "author_flair_text": null,
        "quarantine": false,
        "title": "Best React Tutorials and Courses",
        "created_utc": 1487558915.0,
        "distinguished": null,
        "media": null,
        "num_comments": 2,
        "visited": false,
        "subreddit_type": "public",
        "ups": 12
      }
    }],
    "after": "t3_5uvujz",
    "before": null
  }
}
