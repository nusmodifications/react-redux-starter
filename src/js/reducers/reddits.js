// @flow
import type { FSA } from 'types/redux';

import { FETCH_REDDITS } from 'actions/reddits';
import * as RequestResultCases from 'middlewares/requests-middleware';

export type RedditsThread = {
  id: string,
  title: string,
  permalink: string,
};
export type RedditsState = Array<{
  data: RedditsThread,
  kind: string,
}>;
export const defaultRedditsState: RedditsState = [];

function reddits(state: RedditsState = defaultRedditsState, action: FSA) {
  switch (action.type) {
    case FETCH_REDDITS + RequestResultCases.SUCCESS:
      return action.payload.data.children;
    default:
      return state;
  }
}

export default reddits;
