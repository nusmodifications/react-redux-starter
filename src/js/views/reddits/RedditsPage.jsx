// @flow
import type { StoreState } from 'reducers';
import type { RequestStates } from 'types/redux';
import type { RedditsState } from 'reducers/reddits';

import React, { Component } from 'react';
import DocumentTitle from 'react-document-title';
import autoBind from 'react-autobind';
import { connect } from 'react-redux';
import { fetchReddits } from 'actions/reddits';

type Props = {
  items: RedditsState,
  fetchReddits: Function,
  fetchRedditsRequest: RequestStates,
};

type State = {
  topic: string,
};

export class RedditsPage extends Component {
  constructor(props: Props) {
    super(props);
    this.state = {
      topic: 'reactjs',
    };
    autoBind(this);
  }

  state: State;

  componentDidMount() {
    this.search();
  }

  props: Props;

  search() {
    this.props.fetchReddits(this.state.topic);
  }

  render() {
    return (
      <DocumentTitle title="Reddit">
        <div className="reddits-page">
          <h1>Reddit</h1>
          <hr/>
          <p>Try searching for &#34;redux&#34; to see the failure case.</p>
          <form onSubmit={(event) => {
            event.preventDefault();
            this.search();
          }}
          >
            <input className="form-control" value={this.state.topic} onChange={(event) => {
              this.setState({
                topic: event.target.value,
              });
            }}
            />
          </form>
          <br/>
          {this.props.fetchRedditsRequest.isPending &&
            <div className="reddits-loading">Loading...</div>
          }
          {this.props.fetchRedditsRequest.isFailure &&
            <div className="reddits-failed">Request failed</div>
          }
          {this.props.fetchRedditsRequest.isSuccessful &&
            <div className="reddits-loaded">
              <ul>
                {this.props.items.map((item) => {
                  return (
                    <li key={item.data.id}>
                      <a href={`https://reddits.com${item.data.permalink}`}
                        target="_blank"
                        rel="noopener noreferrer"
                        title={item.data.title}
                      >
                        {item.data.title}
                      </a>
                    </li>
                  );
                })}
              </ul>
            </div>
          }
        </div>
      </DocumentTitle>
    );
  }
}

export function mapStateToProps(state: StoreState) {
  return {
    items: state.reddits,
    fetchRedditsRequest: state.requests.fetchRedditsRequest || {},
  };
}

export default connect(
  mapStateToProps,
  {
    fetchReddits,
  },
)(RedditsPage);
