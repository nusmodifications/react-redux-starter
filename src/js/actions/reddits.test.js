import * as redditsActions from './reddits';

describe('reddits actions', () => {
  it('should create an action to fetch reddits', () => {
    const action: FSA = redditsActions.fetchReddits('redux');
    expect(action).toMatchSnapshot();
  });
});
