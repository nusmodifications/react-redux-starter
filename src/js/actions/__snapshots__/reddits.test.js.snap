// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reddits actions should create an action to fetch reddits 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "https://www.reddit.com/r/redux.json",
  },
  "type": "FETCH_REDDITS",
}
`;
